### PR TITLE
feat(RELEASE-400): adds check-data-keys task in release pipeline

### PR DESCRIPTION
**Pull request by seanconroy2021**
This commit adds the check-data-keys task into the release pipelines to ensure validation of required data keys.
[Link](https://issues.redhat.com/browse/RELEASE-400)

### CODE REVIEW

File: schema/dataKeys.json
Comment (seanconroy2021) [inline schema/dataKeys.json:53]: `  /workspace/data/5c52c6c8-898a-4a85-93b2-9336e6601e52/data.json::$.fbc.productName: 'testProductName' does not match '^[a-z]+$'`
  
  Just removing for time being to get E2E to pass.... I would say we can remove it

File: pipelines/rh-advisories/rh-advisories.yaml
Comment (johnbieren) [inline pipelines/rh-advisories/rh-advisories.yaml:1]: This is the same as `rh-push-to-registry-redhat-io` but + advisories, so I think it should also have the systems that one has

File: pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml
Comment (johnbieren) [inline pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: I don't think this is required

File: pipelines/rhtap-service-push/rhtap-service-push.yaml
Comment (johnbieren) [inline pipelines/rhtap-service-push/rhtap-service-push.yaml]: I don't see signing tasks in this pipeline, but I do see slack stuff

**Review by johnbieren — COMMENTED** [review]
E2E results may be interesting on this one :joy:

File: schema/dataKeys.json
Comment (mmalina) [reply schema/dataKeys.json:53]: Or maybe just relax it to allow capital letters? Or modify e2e to contain lowercase in the name? That should be easy. Maybe just this here? https://github.com/konflux-ci/e2e-tests/blob/fd26460f5d0035b0bffe699513c558be8588b209/tests/release/pipelines/fbc_release.go#L347

File: pipelines/fbc-release/README.md
Comment (mmalina) [inline pipelines/fbc-release/README.md]: I would suggest to use the verb "Add" instead of "Implement"  - because that's what being done, right? We're adding the new task. (This applies to all the others as well.)

File: pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml
Comment (mmalina) [inline pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: In rh-advisories, this has verify-enterprise-contract in runAfter. Not sure which one is more correct, but they should do the same - and if run after verify-enterprise-contract is correct, then perhaps it should be used in other pipelines that have it as well.

File: pipelines/fbc-release/README.md
Comment (seanconroy2021) [reply pipelines/fbc-release/README.md]: Yeah, we are adding .. will change it  
Fixed

File: schema/dataKeys.json
Comment (mmalina) [inline schema/dataKeys.json:429]: What's the motivation for this change? I don't know what's allowed for `contentGateway`, but I guess there was a reason to not allow additional properties?

Comment (seanconroy2021) [reply schema/dataKeys.json:429]: It was down to extra params being added that are not user provided).
[Slack Thread here](https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1731934464613269?thread_ts=1731665042.843589&cid=C04PZ7H0VA8)

Comment (mmalina) [reply schema/dataKeys.json:429]: Ah, ok, so it was based on a discussion, then it's good.

File: pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml
Comment (seanconroy2021) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: They all should be run after `collect-data`  ... Will update the `push-to-cdn`
done

Comment (mmalina) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: My comment is about rh-advisories ;) Not sure what you mean by "will update push-to-cdn" - that's a task. Which pipeline are you talking about?

Comment (seanconroy2021) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: I'm confused as ` rh-advisories ` has populate-release-notes-images in its run after for check-data-leys.

File: pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml
Comment (seanconroy2021) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: Fixed

File: schema/dataKeys.json
Comment (johnbieren) [reply schema/dataKeys.json:53]: I hit this with some of the advisory keys too. I am fine with either removing it (the pattern, not the description) or modifying like Martin said so long as it is not too restricting such that users are going to fail on this

File: pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml
Comment (johnbieren) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: Yeah, and we need to keep it that way AFAIK because it checks some of the keys the populate-release-notes-images task adds

Comment (mmalina) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: > we need to keep it that way AFAIK because it checks some of the keys the populate-release-notes-images task adds

You mean if we checked the keys after populate-release-notes-images, it might not pass? But we just discussed elsewhere that the schema does include the releaseNotes.content.images, right? So it should be fine?

But I'm not sure who was even proposing we do the check after populate-release-notes-images here, certainly not me. 

My original question still stands: rh-advisories has verify-enterprise-contract in runAfter. Do we want that, or do we want collect-data? (Of course if we decide to have it after populate-release-notes-images in rh-advisories, this question would no longer be relevant.)

Comment (johnbieren) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: rh-advisories does not have that. Are you looking at the right pipeline? It has this
```
    - name: check-data-keys
      params:
        - name: dataPath
          value: "$(tasks.collect-data.results.data)"
        - name: schema
          value:  $(params.taskGitUrl)/raw/$(params.taskGitRevision)/schema/dataKeys.json
        - name: systems
          value:
            - releaseNotes
      taskRef:
        params:
          - name: url
            value: $(params.taskGitUrl)
          - name: revision
            value: $(params.taskGitRevision)
          - name: pathInRepo
            value: tasks/check-data-keys/check-data-keys.yaml
        resolver: git
      workspaces:
        - name: data
          workspace: release-workspace
      runAfter:
        - populate-release-notes-images
```

File: pipelines/rhtap-service-push/rhtap-service-push.yaml
Comment (seanconroy2021) [reply pipelines/rhtap-service-push/rhtap-service-push.yaml]: I have removed the signing. I thought Slack is not required  [Slack](https://github.com/seanconroy2021/release-service-catalog/blob/003c505a2127ee4f506736d0b8d9eb9435199ff9/tasks/send-slack-notification/send-slack-notification.yaml#L53-L54). I could be wrong

File: pipelines/rh-advisories/rh-advisories.yaml
Comment (seanconroy2021) [reply pipelines/rh-advisories/rh-advisories.yaml:1]: Added in mapping, pyxis and sign into `rh-advisories`

File: schema/dataKeys.json
Comment (seanconroy2021) [reply schema/dataKeys.json:53]: Ok so I have allowed for the pattern to have capital letters

File: pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml
Comment (mmalina) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: > Are you looking at the right pipeline?

Thanks. I'm not sure where I was looking. You're right it's not there. Sorry about the noise.

But I do see it in `push-disk-images-to-cdn` pipeline:
```
    - name: check-data-keys
      params:
        - name: dataPath
          value: "$(tasks.collect-data.results.data)"
        - name: systems
          value:
            - cdn
      taskRef:
        params:
          - name: url
            value: $(params.taskGitUrl)
          - name: revision
            value: $(params.taskGitRevision)
          - name: pathInRepo
            value: tasks/check-data-keys/check-data-keys.yaml
        resolver: git
      workspaces:
        - name: data
          workspace: release-workspace
      runAfter:
        - verify-enterprise-contract

```

Do we know why? Shouldn't it also run right after collect-data there?

File: schema/dataKeys.json
Comment (mmalina) [reply schema/dataKeys.json:53]: looks good to me

Comment (seanconroy2021) [reply schema/dataKeys.json:53]: We also get this one `.fbc.productName: 'preGA-product' does not match '^[a-zA-Z]+$.'` I could add in to allow hyphen or just update e2e.

File: pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml
Comment (seanconroy2021) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: I would imagine it should run right after the collect-data like the rest.

File: schema/dataKeys.json
Comment (johnbieren) [reply schema/dataKeys.json:53]: Up to you. I don't know that we have the exact requirements on this field which is why I say we can just remove it, but no strong opinion

File: pipelines/rhtap-service-push/rhtap-service-push.yaml
Comment (johnbieren) [reply pipelines/rhtap-service-push/rhtap-service-push.yaml]: You're right

**Review by johnbieren — DISMISSED** [review]
lgtm, would be nice to get Leandro's input on the fbc keys before merging

File: pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml
Comment (mmalina) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: Yeah, I'd do that.

Comment (seanconroy2021) [reply pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml]: Changed it

**Review by johnbieren — DISMISSED** [review]

**Review by mmalina — DISMISSED** [review]

**Comment by johnbieren** [comment]
```
                        "lastTransitionTime": "2024-12-10T15:03:18Z",
                        "message": "the step \"reduce\" in TaskRun \"managed-gld9m-reduce-snapshot\" failed to pull the image \"\". The pod errored with the message: \"Back-off pulling image \"quay.io/enterprise-contract/ec-cli@sha256:913c7dac3d41877b01835d2e55bcd970c6cdbf4944f8176e9e3de9548642a2b4\".\"",
                        "reason": "TaskRunImagePullFailed",
                        "status": "False",
                        "type": "Succeeded"
```

File: pipelines/push-disk-images-to-cdn/push-disk-images-to-cdn.yaml
Comment (theflockers) [inline pipelines/push-disk-images-to-cdn/push-disk-images-to-cdn.yaml]: If I am not mistaken, for new parameters you should update the minor  - or 1.3.0 instead.

File: schema/dataKeys.json
Comment (theflockers) [inline schema/dataKeys.json]: we have 3 types of FBC release: normal, preGa and hotfix. Some are common, other are only for preGA and other for hotfix. If I understand correctly, would this make all FBC require these keys?

File: tasks/check-data-keys/check-data-keys.yaml
Comment (theflockers) [inline tasks/check-data-keys/check-data-keys.yaml]: adding a -L requires only bumping the patch, so 1.0.1.

File: schema/dataKeys.json
Comment (seanconroy2021) [reply schema/dataKeys.json]: So this would make all of them required. Which, I am guessing, would not work then?

File: pipelines/push-disk-images-to-cdn/push-disk-images-to-cdn.yaml
Comment (seanconroy2021) [reply pipelines/push-disk-images-to-cdn/push-disk-images-to-cdn.yaml]: Fixed

File: tasks/check-data-keys/check-data-keys.yaml
Comment (seanconroy2021) [reply tasks/check-data-keys/check-data-keys.yaml]: Fixed

File: schema/dataKeys.json
Comment (theflockers) [reply schema/dataKeys.json]: Yes. Unless there is a way to create some conditions, like if hotfix, then... if preGA, then... otherwise, we should make only the common tags required. Also, I forgot to mention that when "stagedIndex" is true, the targetIndex should be empty :D

Comment (theflockers) [reply schema/dataKeys.json]: The following is true:

Required *common* parameters:
```
      fromIndex
      requestUpdateTimeout 
      buildTimeoutSeconds
      publishingCredentials
      allowedPackages
```
required for *prod*  builds
```
all common
      targetIndex
```
required for *stagedIndex* builds
```
all common
      stagedIndex

``` 
required for *hotfix* builds
```
all common
      targetIndex
      hotfix
      issueId

```
 required for *preGA* builds
```
all common
      targetIndex
      preGA
      productName
      productVersion
```

Comment (seanconroy2021) [reply schema/dataKeys.json]: Thanks for the breakdown, I should be able to make it work

Comment (theflockers) [reply schema/dataKeys.json]: I updated the comment as the targetIndex was missing for preGA.

Comment (seanconroy2021) [reply schema/dataKeys.json]: Ok, so i will able to make preGA and hotfix work from using if preGA or hotfix is true to make .. required. I cant see a way to make the others required since they only have one key

File: tasks/check-data-keys/README.md
Comment (johnbieren) [inline tasks/check-data-keys/README.md]: Since you're fixing the default, can you make it what it actually is in the yaml? `https://raw.githubusercontent.com/konflux-ci/release-service-catalog/refs/heads/development/schema/dataKeys.json`

Comment (seanconroy2021) [reply tasks/check-data-keys/README.md]: will do fixed

**Review by johnbieren — DISMISSED** [review]
lgtm, but let Leandro approve the fbc changes before merging please

File: schema/dataKeys.json
Comment (theflockers) [reply schema/dataKeys.json]: I think this is fine. The two you mentioned, their functionality rely on having these fields set. In the other hand, the stagedIndex will work the same way having the targetIndex or not.

Comment (theflockers) [reply schema/dataKeys.json]: Ah, the prod build means that stagedIndex is not set. Maybe that; s the only thing that still worth it to check. If `stageIndex` is false (or unset), the targetIndex is required.

Comment (seanconroy2021) [reply schema/dataKeys.json]: Fixed

**Review by theflockers — COMMENTED** [review]
LGTM once squashed and the failing test is fixed.

**Review by theflockers — DISMISSED** [review]

**Comment by johnbieren** [comment]
@seanconroy2021 are you looking into why it fails or just retesting? E2E has been relatively stable this week

**Comment by seanconroy2021** [comment]
> @seanconroy2021 are you looking into why it fails or just retesting? E2E has been relatively stable this week

I have been looking at the logs not getting any info on why it failing... it keeps passing in the evening

**Comment by johnbieren** [comment]
> > @seanconroy2021 are you looking into why it fails or just retesting? E2E has been relatively stable this week
> 
> I have been looking at the logs not getting any info on why it failing... it keeps passing in the evening

If it is sometimes passing, you can enable auto-merge so that it will merge as soon as it passes e2e (assuming it is rebased)

**Comment by seanconroy2021** [comment]
> > > @seanconroy2021 are you looking into why it fails or just retesting? E2E has been relatively stable this week
> > 
> > 
> > I have been looking at the logs not getting any info on why it failing... it keeps passing in the evening
> 
> If it is sometimes passing, you can enable auto-merge so that it will merge as soon as it passes e2e (assuming it is rebased)

Ok rebased, can you give approval again so it will auto-merge

**Review by johnbieren — DISMISSED** [review]

**Review by johnbieren — APPROVED** [review]

**Pull request merged**
